[BLOG] Amazon Q CLI Integration: Log Pattern and Data Distribution Analysis (#3984)
opensearch-project/project-website: _posts/2025-10-17-Q-CLI-Integration:-Log-Pattern-and-Data-Distribution-Analysis.md

---
layout: post
title: "Q CLI Integration: Log Pattern and Data Distribution Analysis"
authors:
  - ihailong
  - jiaruj
date: 2025-10-17
categories:
  - technical-posts

> meta_keywords: Q CLI, CLI integration, log pattern analysis, data distribution analysis, OpenSearch CLI, command line interface, log analysis tools, data distribution tools, OpenSearch integration, diagnostic tools

Reviewer: Add MCP, Agent, Agentic AI, OpenTelemetry as keyword

meta_description: Learn how to integrate Q CLI with OpenSearch's log pattern analysis and data distribution tools for streamlined command-line diagnostics and enhanced system troubleshooting capabilities.
---

## Log Pattern Analysis Tool Overview

> The [Log Pattern Analysis Tool](https://docs.opensearch.org/latest/ml-commons-plugin/agents-tools/tools/log-pattern-analysis-tool/) is an advanced OpenSearch agent tool that automates root cause analysis through multiple analysis modes. It performs differential pattern analysis between baseline and problem periods, analyzes log sequences using trace correlation, and provides log insights by automatically extracting error patterns and keywords from log data to accelerate troubleshooting.

Reviewer: Advanced can be removed

## Data Distribution Tool Overview

The [Data Distribution Tool](https://docs.opensearch.org/latest/ml-commons-plugin/agents-tools/tools/data-distribution-tool/) is an OpenSearch agent tool that analyzes field value distributions within your indices. It generates statistical summaries including value frequencies, percentiles, and distribution metrics to help understand data characteristics and identify potential data quality issues.

## Q CLI Integration

> Q CLI can seamlessly integrate with these OpenSearch analysis tools through the [Model Context Protocol (MCP)](https://opensearch.org/blog/introducing-mcp-in-opensearch/). By configuring the OpenSearch MCP server, you can access both the Log Pattern Analysis Tool and Data Distribution Tool directly from your command line interface. This integration enables natural language queries for log analysis and data distribution insights, making complex diagnostic tasks accessible through simple conversational commands.

Author: Thanks for your suggestion. I've resolved it.

### Q CLI Investigation Process

Using Q CLI with the integrated OpenSearch tools, we can investigate this issue through natural language queries. Here's how the investigation unfolds:

**Initial Query:**
```
Users are reporting payment failures during checkout process and we need to investigate the root cause by analyzing logs from the ss4o_logs* index between 2025-08-11 07:29:07.414 and 2025-08-12 07:29:07.414 to identify patterns in checkout service failures related to card charging issues.
```

**Q CLI Analysis Results:**
```
I'll help you investigate the payment failures by analyzing the logs in the ss4o_logs* index. Let me start by examining the index structure and then search for checkout-related failures.
> 
> 🛠️  Using tool: ListIndexTool from mcp server opensearch
>  ⋮ 
>  ● Running ListIndexTool with the param:
>  ⋮  {
>  ⋮    "index": "ss4o_logs"
>  ⋮  }
> 
>  ⋮ 
>  ● Completed in 1.187s
> 
> > Let me check what indices are available that match the pattern:
> 
> 🛠️  Using tool: ListIndexTool (trusted) from mcp server opensearch
>  ⋮ 

> >  ● Running ListIndexTool with the param:

Author: This is a concrete example from my OpenTelemetry testing; I pasted the full input here. It does look incorrect, so I’ll delete these lines.

### Implementation with opensearch-mcp-server-py

This integration is built on the [opensearch-mcp-server-py](https://github.com/opensearch-project/opensearch-mcp-server-py) project, which provides a Python-based MCP server for OpenSearch. To enable the Log Pattern Analysis Tool and Data Distribution Tool integration, you need to clone this project and add custom integration code for both tools, extending the server's capabilities to support these advanced analysis features.

### Complete Integration Workflow

To set up Amazon Q CLI with OpenSearch MCP integration, follow these steps:

1. **Clone the MCP Server Repository**
   ```bash
   git clone https://github.com/opensearch-project/opensearch-mcp-server-py.git
   cd opensearch-mcp-server-py
   ```

2. **Add Tool Integration Code**

>    - Implement Log Pattern Analysis Tool integration in the MCP server

Author: I forked the mcp-server repository and pushed my test code there. I’ve added the GitHub link in the blog.

   - Automatically grouped related error messages into logical failure categories.
   - Provided exact occurrence count statistics.
   - Delivered actual log samples for each pattern for validation.

**Value delivered**: This tool eliminated the need for manual pattern recognition, automatically discovering that payment token validation failures occur 3.3 times more frequently than product catalog issues, clearly establishing primary and secondary failure mode priorities.

#### Data Distribution tool's contribution

The Data Distribution tool contributed to identifying the root cause of payment failures in the following ways.

**Core contribution: Statistical analysis and context provision**

The Data Distribution tool provided critical statistical background and field distribution analysis for the investigation:

1. **Service distribution analysis** (divergence: 0.223):
   - `kafka`: 22% (highest log volume service)
   - `product-catalog`: 18% (secondary failure source)
   - `frontend-proxy`: 15% (user-facing errors)
   - `checkout`: 13% (primary failure point)
   - `load-generator`: 11%
   - Other services: 21%

2. **Severity level analysis** (divergence: 0.609):

>    - INFO level: 61% (normal operations)

Reviewer 2: It's not clear why there are different info levels. I'm assuming this is because the demo simulates different services where you have both java "INFO" and python "info" and maybe .net "Information"? If so, please add this in parentheses. If not, please add an explanatory sentence so readers can understand the discrepancy. Put actual error code levels in code font (so `error` or `INFO`).

Author: Thanks for your suggestion. I've resolved it.

---
layout: post
title: "Amazon Q CLI integration: Log pattern and data distribution analysis"
authors:
  - ihailong
  - jiaruj

> date: 2025-10-17

Reviewer 3: Update: 2025-11-20

categories:
  - technical-posts
meta_keywords: Amazon Q CLI, CLI integration, log pattern analysis, data distribution analysis, OpenSearch CLI, command line interface, log analysis tools, data distribution tools, OpenSearch integration, diagnostic tools, MCP, Agent, Agentic AI, OpenTelemetry

> meta_description: Learn how to integrate Amazon Q CLI with OpenSearch's log pattern analysis and data distribution tools for streamlined command-line diagnostics and enhanced system troubleshooting capabilities.

Author: Thanks for your suggestion. I've resolved it.

Through a real-world OpenTelemetry Demo scenario investigating payment failures, you'll learn how to set up the integration, configure the necessary components, and use these tools to perform automated pattern recognition and statistical analysis. By the end of this post, you'll understand how to use conversational commands to quickly identify root causes in distributed system logs, significantly reducing time to resolution for critical issues.

## Log Pattern Analysis tool

The [Log Pattern Analysis Tool](https://docs.opensearch.org/latest/ml-commons-plugin/agents-tools/tools/log-pattern-analysis-tool/) is an OpenSearch agent tool that automates log analysis through multiple analysis modes. It performs differential pattern analysis between baseline and problem periods and analyzes log sequences using trace correlation. The tool provides log insights by automatically extracting error patterns and keywords from log data to accelerate troubleshooting.

## Data Distribution tool

> The [Data Distribution Tool](https://docs.opensearch.org/latest/ml-commons-plugin/agents-tools/tools/data-distribution-tool/) is an OpenSearch agent tool that analyzes data distribution patterns within datasets and compares distributions between different time periods. It supports both single dataset analysis and comparative analysis to identify significant changes in field value distributions, helping detect anomalies, trends, and data quality issues.

Reviewer 4: ```suggestion
The [Data Distribution tool](https://docs.opensearch.org/latest/ml-commons-plugin/agents-tools/tools/data-distribution-tool/) is an OpenSearch agent tool that analyzes data distribution patterns within datasets and compares distributions between different time periods. It supports both single dataset analysis and comparative analysis to identify significant changes in field value distributions, helping detect anomalies, trends, and data quality issues.
```

In this environment, common issues arise such as payment processing failures, cart abandonment errors, recommendation service timeouts, and checkout workflow disruptions. This provides excellent opportunities for root cause analysis in a distributed system environment.

### Investigating payment failures

In our scenario, users are reporting payment failures during checkout---a critical issue impacting revenue and customer experience. We choose payment failures as our root cause analysis target due to their business-critical impact.

The OpenTelemetry demo can simulate various failure conditions, including payment service disruptions and authentication issues. To resolve this issue, we need to analyze the logs to identify failure patterns and determine whether the issue is related to specific customer segments, authentication problems, or system-level configuration issues.

### Amazon Q CLI investigation process

> Using Amazon Q CLI with the integrated OpenSearch tools, you can investigate this issue through natural language queries. Here's how the investigation unfolds:

Reviewer 4: ```suggestion
Using Amazon Q CLI with the integrated OpenSearch tools, you can investigate this issue through natural language queries. Here's how the investigation unfolds.
```

                    "selectionPercentage": 0.13
                  },
                  {
                    "value": "load-generator",
                    "selectionPercentage": 0.11
                  },
                  {
                    "value": "cart",
                    "selectionPercentage": 0.08
                  },
                  {
                    "value": "shipping",
                    "selectionPercentage": 0.04
                  },
                  {
                    "value": "ad",
                    "selectionPercentage": 0.03
                  },
                  {
                    "value": "currency",
                    "selectionPercentage": 0.02
                  },
                  {
                    "value": "recommendation",
                    "selectionPercentage": 0.01
                  }
                ]
              }
            ]
          }
        }
      ]
    }
  ]
}
```

#### Log Pattern Analysis tool's contribution

The Log Pattern Analysis tool contributed to identifying the root cause of payment failures in the following ways.

**Core contribution: Automated error pattern recognition**

> The Log Pattern Analysis Tool played a crucial pattern identification role in this payment failure investigation:

Reviewer 4: ```suggestion
The Log Pattern Analysis tool played a crucial pattern identification role in this payment failure investigation:
```

   - Empty values: 18%

3. **Field anomaly detection**:
   - `droppedAttributesCount`: divergence = 1.0 (complete anomaly)
   - `severityNumber`: divergence = 0.689 (high anomaly)
   - `flags`: divergence = 0.594 (moderate anomaly)
   - `schemaUrl`: divergence = 0.573 (moderate anomaly)

**Value delivered**: This tool revealed that despite checkout service representing only 13% of total logs, it contains the highest concentration of critical failures. The severity distribution showed that error-level logs are rare (1%), making the 63 payment failures statistically significant. This quantitative context helped prioritize checkout service investigation over higher-volume but less critical services like `kafka`.

#### Both tools working together

The combination of both tools created a synergistic effect that enhanced the investigation's effectiveness:

1. **Qualitative + quantitative analysis**: Log Pattern Analysis tool provided specific error patterns, Data Distribution tool provided statistical validation.
2. **Priority guidance**: Combined analysis showed checkout service has disproportionately high failure impact despite lower log volume.

> 3. **Root cause validation**: Both tools confirmed payment token validation as the primary issue, with product catalog as secondary.

Reviewer 4: ```suggestion
3. **Root cause validation**: Both tools confirmed payment token validation as the primary issue, with the product catalog as secondary.
```

> 4. **Actionable insights**: Tools working together provided specific error messages and statistical significance, supporting clear remediation recommendations.

Reviewer 4: ```suggestion
4. **Actionable insights**: The tools worked together to provide specific error messages and statistical significance, supporting clear remediation recommendations.
```
